### PR TITLE
Enable bigger, more realistic toy matrices (SCP-2464)

### PR DESCRIPTION
**Pull request by eweitz**
This enables `make_toy_data.py` to reliably generate toy dense expression matrices with 1000's of genes.  Toy matrices now also use realistic 15-digit expression values instead of 2- and 3-digit values.  Such test data will help prove (and improve) Ingest Pipeline performance.

Previously, `make_toy_data.py` fetched human gene names for the toy matrix from the NCBI E-Utils HTTP API.  Rate limits were imposed by that API since this code was first written.  So instead of throttling API requests -- which would make the slow process of fetching 10K+ gene names even slower -- `make_toy_data.py` now parses gene names from an [Ensembl GTF](http://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/).  This GTF is quickly and easily retrieved (and cached) by reusing code from [`/bin/python/genomes` in the single_cell_portal_core repo](https://github.com/broadinstitute/single_cell_portal_core/tree/7a734f8982f5891a5fe588db4c746ce6ef22d00c/bin/python/genomes).

Reusing that code makes fetching genes to generate realistic matrices > 100x faster than before.  

The reused code is in `/scripts/genomes` in this repo.  This repo's copy has been modified to enable module import and to better parameterize e.g. species.  While expedient (and discussed with Devon), having slightly different copies of non-trivial code in two repos is _high interest tech debt_.  SCP-2470 outlines how to consolidate the code and pay off the tech debt.

This enhanced code was used to make realistic toy dense matrices -- one 1M cell by 10K gene matrix to test upper limits of Ingest Pipeline and several smaller matrices to evaluate its computational complexity in run-time and space (i.e. consumed RAM).

You can reproduce a 1M by 10K toy dense matrix like so (beware, it needs ~62 GB and > 4 hours):
```
python3 make_toy_data.py --num-files=1 --num-cells 1_000_000 --num-genes 10_000
```

You can find the matrices in [gs://broad-singlecellportal-staging-testing-data/toy_dense_data/](https://console.cloud.google.com/storage/browser/broad-singlecellportal-staging-testing-data/toy_dense_data;tab=objects?project=broad-singlecellportal-staging).

The main code for this review is `make_toy_data.py`.  Code in `/scripts/genomes` would benefit from close review once SCP-2470 is ready.

This satisfies SCP-2464.

### CODE REVIEW

File: scripts/genomes/parse_genome_annotations.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        dir = '/pub/release-' + release + '/gtf/' + organism + '/'
+        filename = organism_upper + '.' + assembly + '.' + release + '.gtf.gz'
+
+        gtf_url = origin + dir + filename
+        gtf_urls.append(gtf_url)
+        ensembl_metadata[taxid]['gtf_path'] = output_dir + filename
+
+    return [gtf_urls, ensembl_metadata]
+
+def transform_ensembl_gtf(gtf_path, ref_dir):
```
Comment (bistline) [inline scripts/genomes/parse_genome_annotations.py:73]: My only issue with this method is that it relies on two utilities from `HTSlib`, but that isn't declared as a dependency anywhere.

File: scripts/genomes/parse_genome_annotations.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    parser.add_argument('--input-dir',
+                        help='Input directory; where to find organisms.tsv.  Default: ./',
+                        default='./')
+    parser.add_argument('--local-output-dir',
+                        help='Local directory for output.  Default: output/',
+                        default='output/')
+    parser.add_argument('--gcs-bucket',
+                        help='Name of GCS bucket for upload.  ' +
+                        'Default: reference_data_dev/',
+                        default='fc-bcc55e6c-bec3-4b2e-9fb2-5e1526ddfcd2')
```
[the remark this answers is not in the document]
Comment (eweitz) [reply scripts/genomes/parse_genome_annotations.py:187]: Agreed, this is a pre-existing issue in the code that was copied.  

I just added an AC in SCP-2470 to follow up on this, while addressing the larger tech debt represented by the duplicate code in `/scripts/genomes`.

File: scripts/genomes/parse_genome_annotations.py (hunk 1 above)
Comment (eweitz) [reply scripts/genomes/parse_genome_annotations.py:73]: Good spot.  This non-Python dependency will likely need a README entry over in `scp-ingest-pipeline` and some informative error handling.  I'll plan to do that while consolidating this duplicate code in `/scripts/genome`, per SCP-2470.  I added this an AC to that ticket.

File: scripts/make_toy_data.py [hunk 3]
```diff
@@ -82,19 +90,13 @@
...
+        'Number of genes (rows)'
     )
 )
 args.add_argument(
     '--num-cells', default=None, type=int,
     help=(
-        'Number of cells'
-    )
-)
-args.add_argument(
```
Comment (jlchang) [inline scripts/make_toy_data.py:94]: I appreciate the performance improvement and increased reliability of parsing a fetched copy of an Ensembl GTF but also advocate for keeping the pre-loaded genes option. 

Two possible uses for pre-loaded genes:
1. need to generate a matrix that has more genes than the most current human Ensemble annotation. (make_toy_data.py currently would limit the generated toy matrix to the number of genes in the fetched human Ensemble annotation).
2. need to simulate/replicate visualization with a specific list of gene names (eg. non-human gene names or an expression matrix for nonRNAseq data, scATAC-seq "gene names" would be genome coordinate ranges) for troubleshooting

File: scripts/make_toy_data.py [hunk 4]
```diff
@@ -235,33 +218,38 @@ def fetch_cells(prefix):
...
     if preloaded_barcodes:
         with open(preloaded_barcodes) as f:
             # load preloaded barcodes/cell names
             lines = f.readlines()
             barcodes = [line.strip() for line in lines if len(line) > 2]
             if num_columns > len(barcodes):
-                # if user param --num_barcdes is higher than the number in the preloaded file, drop it down
+                # if user param --num-barcodes is higher than the number in the
+                # preloaded file, drop it down
                 print('Not enough barcodes in preloaded file, reducing gene number to', len(genes))
```
Comment (jlchang) [inline scripts/make_toy_data.py:240]: The content of this print statement seems inaccurate
```suggestion
                print('Not enough barcodes in preloaded file, reducing barcode number to', len(barcodes))
```

File: scripts/make_toy_data.py (hunk 3 above)
Comment (eweitz) [reply scripts/make_toy_data.py:94]: Helpful context, I've restored `--preloaded-genes` (5474edd).  I've also noted those use cases in the `--help` docs for that argument (0646433).

File: scripts/make_toy_data.py (hunk 4 above)
Comment (eweitz) [reply scripts/make_toy_data.py]: Nice catch!  Fixed in 5474edd.

**Pull request merged**